ddtrace/internal/coverage/lines.py

```diff
@@ -0,0 +1,80 @@
+import typing as t
+
+
+def _bit_count_compat(_byte: int) -> int:
+    return bin(_byte).count("1")
+
+
+_bit_count: t.Callable[[int], int] = int.bit_count if hasattr(int, "bit_count") else _bit_count_compat
+
+
+class CoverageLines:
+    def __init__(self, initial_size: int = 32):
+        # Initial size of 32 chosen based on p50 length of files in code base being 240 at time of writing
+        self._lines = bytearray(initial_size)
+
+    def __len__(self):
+        return self._num_lines()
+
+    def __bool__(self):
+        return self._num_lines() > 0
+
+    def __copy__(self):
+        new_instance = CoverageLines()
+        new_instance._lines = self._lines.copy()
+        return new_instance
+
+    def __repr__(self):
+        return f"CoverageLines(num_lines={self._num_lines()})"
+
+    def _num_lines(self) -> int:
+        return sum(_bit_count(byte) for byte in self._lines)
+
+    def add(self, line_number: int):
+        lines_byte = line_number // 8
+
+        if lines_byte >= len(self._lines):
+            self._lines.extend(bytearray(lines_byte - len(self._lines) + 1))
+
+        # DEV this fun bit allows us to trick ourselves into little-endianness, which is what the backend wants to see
+        # in bytes
+        lines_bit = 2 ** (7 - (line_number % 8))
+        self._lines[lines_byte] |= lines_bit
+
+    def add_lines(self, lines: list[int]):
+        for line in lines:
+            self.add(line)
+
+    def to_bytes_string(self) -> str:
+        return " ".join(bin(byte)[2:].zfill(8) for byte in self._lines)
```

Comment:
We can also do this, but I don't know if it's an improvement:
```suggestion
        return " ".join(f"{byte:08b}" for byte in self._lines)
```

Reply:
Resolving because I'm removing this.